src/main/java/net/minecraftforge/client/gui/config/CategoryConfigScreen.java

```diff
@@ -0,0 +1,294 @@
+import java.util.*;
+import java.util.function.BooleanSupplier;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link ConfigScreen} for a (sub) category.
+ * A category is made when you do "builder.push()" when initialising your config.
+ *
+ * @author Cadiboo
+ */
+public class CategoryConfigScreen extends ConfigScreen {
+    public static final TextFormatting TOOLTIP_TITLE_COLOR = TextFormatting.GREEN;
+    public static final TextFormatting TOOLTIP_COMMENT_COLOR = TextFormatting.YELLOW;
+    public static final TextFormatting TOOLTIP_EXTRA_DATA_COLOR = TextFormatting.AQUA;
+    public static final TextFormatting TOOLTIP_WORLD_RESTART_REQUIRED_COLOR = TextFormatting.RED;
+    protected final ConfigCategoryInfo categoryInfo;
+
+    public CategoryConfigScreen(Screen parentScreen, ITextComponent title, ConfigCategoryInfo categoryInfo) {
+        super(parentScreen, title, ConfigScreen.makeSubtitle(parentScreen, title));
+        this.categoryInfo = categoryInfo;
+    }
+
+    static ITextComponent translateWithFallback(@Nullable String translationKey, String fallback) {
+        if (translationKey == null)
+            return new StringTextComponent(fallback);
+        ITextComponent title = new TranslationTextComponent(translationKey);
+        if (translationKey.equals(title.getString()))
+            return new StringTextComponent(fallback);
+        return title;
+    }
+
+    static Collection<? extends ITextProperties> translateCommentWithFallback(@Nullable String translationKey, @Nullable String fallbackComment) {
+        // Need to handle line breaks (\n) properly
+        if (translationKey != null) {
+            TranslationTextComponent comment = new TranslationTextComponent(translationKey);
+            if (!translationKey.equals(comment.getString())) {
+                comment.func_240701_a_(TOOLTIP_COMMENT_COLOR); // mergeStyle
+                return Collections.singletonList(comment);
+            }
+        }
+        if (fallbackComment == null)
+            return Collections.emptyList();
+        return Arrays.stream(fallbackComment.split("\n"))
+            .map(StringTextComponent::new)
+            .map(s -> s.func_240701_a_(TOOLTIP_COMMENT_COLOR)) // mergeStyle
+            .collect(Collectors.toList());
+    }
+
+    public static List<ITextProperties> createTooltip(ITextComponent title, @Nullable String commentTranslationKey, @Nullable String comment) {
+        List<ITextProperties> tooltip = new LinkedList<>();
+        tooltip.add(title.func_230532_e_().func_240701_a_(TOOLTIP_TITLE_COLOR));
+        tooltip.addAll(translateCommentWithFallback(commentTranslationKey, comment));
+        return tooltip;
+    }
+
+    @Override
+    protected ConfigElementList makeConfigElementList() {
+        return new CategoryConfigElementList(this, field_230706_i_, categoryInfo);
+    }
+
+    public interface ConfigCategoryInfo {
```

Comment:
Maybe move the inner classes it into their own files? Or is there a good reasoning this is an inner class?

Reply:
It's a mostly internal thing that is likely change a bit before I merge this PR. I don't see any reason for this to be in it's own class as it's just used like a `Record` to pass around functions. I should probably make this into a POJO rather than an interface though...